src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseDataFileService.cs

```diff
@@ -168,37 +168,32 @@ await ids.Select(id => _releaseFileRepository.CheckFileExists(releaseId, id, Fil
                 .OnSuccess(async file => await GetDataFileInfo(releaseId, file.File));
         }
 
-        public async Task<Either<ActionResult, IEnumerable<DataFileInfo>>> ListAll(Guid releaseId)
+        public async Task<Either<ActionResult, List<DataFileInfo>>> ListAll(Guid releaseId)
         {
             return await _persistenceHelper
                 .CheckEntityExists<Release>(releaseId)
                 .OnSuccess(_userService.CheckCanViewRelease)
-                .OnSuccess(async _ =>
+                .OnSuccess(async () =>
                 {
-                    var fileList = new List<DataFileInfo>();
                     var files = await _releaseFileRepository.GetByFileType(releaseId, FileType.Data);
 
                     // Exclude files that are replacements in progress
-                    var filesExcludingReplacements = files.Where(file =>
-                        !file.File.ReplacingId.HasValue);
+                    var filesExcludingReplacements = files.Where(file => !file.File.ReplacingId.HasValue);
 
-                    await filesExcludingReplacements
+                    return await filesExcludingReplacements
                         .ToAsyncEnumerable()
-                        .ForEachAwaitAsync(async file =>
-                            fileList.Add(await GetDataFileInfo(releaseId, file.File)));
-
-                    return fileList
-                        .OrderBy(file => file.Name)
-                        .AsEnumerable();
+                        .SelectAwait(async file => await GetDataFileInfo(releaseId, file.File))
```

Comment:
Cool, good suggestion. I've made this change in [this commit](https://github.com/dfe-analytical-services/explore-education-statistics/pull/3387/commits/30fe5e924a895555b5327975b542794e6aa86054).

I get the `DataImport` which we were already getting, and then use it to get the data file and the meta file, avoiding the old way of looking up the meta file which associated the two files together by subject id.

If it's a list of files then this happens in one query, although I used `AsSplitQuery` to avoid repetition in the result set.

It works out quite nicely because this change also removes a lookup on the subject to get the subject name which cuts out a bunch of setup in the unit tests.

However it highlights an inefficiency that still exists getting the permissions for each file using a auth handler.

We know the data import status so if we could work out if the user has permission to cancel any import (which Bau users can), then we wouldn't need to make repeated calls to an auth handler for each file.

That's probably not too difficult however I left this as a TODO for another day 🙂.

Reply:
Nice one!